pkg/lib/oauth/handler/handler_token.go

```diff
@@ -2099,7 +2099,9 @@ func (h *TokenHandler) validateClientSecret(client *config.OAuthClientConfig, cl
 }
 
 func (h *TokenHandler) checkRateLimits(ctx context.Context, rl ratelimit.RateLimit, opts ratelimit.ResolveBucketSpecOptions) error {
-	specs := rl.ResolveBucketSpecs(nil, nil, nil, &opts)
+	finalOpts := opts
+	finalOpts.IPAddress = string(h.RemoteIP)
```

Comment:
The IP rate limit should be applied very early in the process pipeline.

If the IP rate limit is applied very late, compute resources are still wasted.